File: Competitions/Aces/aces_coders_v9/Dirty Short.py
t = int(input())

def dfs(n):
    global grid, colour, visited
    
    visited[n] = True
    
    for i in grid[n]:
        
        if not visited[i]:
            colour[i] = 1 if colour[n] == 2 else 2
            if dfs(i) == 1:
                return 1
        else:
            if colour[n] == colour[i]:
                print("YES")
                return 1
        
    return 0

for _ in range(t):
    n, m = [int(x) for x in input().strip().split(" ")]
    
    grid = [set() for x in range(n)]
    colour = [0 for x in range(n)]
    visited = [False for x in range(n)]
    
    for i in range(m):
        a, b = [int(x) for x in input().strip().split(" ")]
        grid[a-1].add(b-1)
        
        
    colour[0] = 1
    if dfs(0) == 0:
        print("NO")

File: Competitions/Aces/aces_coders_v9/Dirty Short.cpp
#include <bits/stdc++.h>
using namespace std;
const int MAX = 2e3 + 1;
vector<vector<int>> grid;
bool visited[MAX];
char colour[MAX];
int t, n, m, a, b;

int dfs(int n)
{

    visited[n] = true;

    for (auto i : grid[n])
    {
        if (visited[i])
        {
            if (colour[i] == colour[n])
            {
                cout << "YES" << endl;
                return 1;
            }
        }
        else
        {
            colour[i] = (colour[n] == 1) ? 2 : 1;
            if (dfs(i) == 1)
                return 1;
        }
    }

    return 0;
}

int main()
{
    cin >> t;

    for (int k = 0; k < t; k++)
    {

        cin >> n >> m;
        grid.clear();
        for (int i = 0; i < n; i++)
        {
            grid.push_back(vector<int>(0));
        }

        for (int i = 0; i < m; i++)
        {
            cin >> a >> b;
            grid[a - 1].push_back(b - 1);
        }

        fill(begin(visited), begin(visited) + n, false);
        fill(begin(colour), begin(colour) + n, 0);

        colour[0] = 1;

        if (dfs(0) == 0)
        {
            if (visited[n - 1])
                cout << "NO" << endl;
            else
                cout << "YES" << endl;
        }
    }

    return 0;
}